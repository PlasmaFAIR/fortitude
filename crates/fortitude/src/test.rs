use std::path::Path;

use anyhow::Result;
use ruff_source_file::{SourceFile, SourceFileBuilder};

use crate::{
    check::{
        ast_entrypoint_map, check_file, read_to_string, rules_to_path_rules, rules_to_text_rules,
    },
    message::{Emitter, TextEmitter},
    rule_table::RuleTable,
    rules::Rule,
    settings::{self, FixMode, Settings},
};

#[macro_export]
macro_rules! apply_common_filters {
    {} => {
        let mut settings = insta::Settings::clone_current();
        // Convert windows paths to Unix Paths.
        settings.add_filter(r"\\\\?([\w\d.])", "/$1");
        let _bound = settings.bind_to_scope();
    }
}

pub(crate) fn test_resource_path(path: impl AsRef<Path>) -> std::path::PathBuf {
    Path::new("./resources/test/").join(path)
}

/// Run [`check_path`] on a Fortran file in the `resources/test/fixtures` directory.
pub(crate) fn test_path(
    path: impl AsRef<Path>,
    rules: &[Rule],
    settings: &Settings,
) -> Result<String> {
    let path = test_resource_path("fixtures").join(path);
    let source = read_to_string(&path)?;
    let filename = path.to_string_lossy();
    let file = SourceFileBuilder::new(filename.as_ref(), source.as_str()).finish();
    Ok(test_contents(&path, &file, rules, settings))
}

pub(crate) fn test_contents(
    path: &Path,
    file: &SourceFile,
    rules: &[Rule],
    settings: &Settings,
) -> String {
    let rule_table = RuleTable::from_iter(rules.iter().cloned());
    let path_rules = rules_to_path_rules(&rule_table);
    let text_rules = rules_to_text_rules(&rule_table);
    let ast_entrypoints = ast_entrypoint_map(&rule_table);

    match check_file(
        &rule_table,
        &path_rules,
        &text_rules,
        &ast_entrypoints,
        path,
        file,
        settings,
        FixMode::Generate,
        settings::IgnoreAllowComments::Disabled,
    ) {
        Ok(violations) => {
            if violations.messages.is_empty() {
                return String::new();
            }

            let mut output = Vec::new();

            TextEmitter::default()
                .with_show_fix_status(true)
                .with_show_fix_diff(true)
                .with_show_source(true)
                .with_unsafe_fixes(crate::settings::UnsafeFixes::Enabled)
                .emit(&mut output, &violations.messages)
                .unwrap();

            String::from_utf8(output).unwrap()
        }
        Err(msg) => {
            panic!("Failed to process: {msg}");
        }
    }
}
