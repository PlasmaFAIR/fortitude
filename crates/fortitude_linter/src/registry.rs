use fortitude_macros::RuleNamespace;
use rayon::str;
use std::str::FromStr; // Needed by strum_macros

pub use crate::rules::Rule;
pub use rule_set::{RuleSet, RuleSetIterator};

mod rule_set;

// Rule categories and identity codes
// ----------------------------------
// Helps to sort rules into logical categories, and acts as a unique tag with which
// users can switch rules on and off.

pub trait AsRule {
    fn rule(&self) -> Rule;
}

impl Rule {
    pub fn from_code(code: &str) -> Result<Self, FromCodeError> {
        let (category, code) = Category::parse_code(code).ok_or(FromCodeError::Unknown)?;
        category
            .all_rules()
            .find(|rule| rule.noqa_code().suffix() == code)
            .ok_or(FromCodeError::Unknown)
    }
}

#[derive(thiserror::Error, Debug)]
pub enum FromCodeError {
    #[error("unknown rule code")]
    Unknown,
}

/// The category of each rule defines the sort of problem it intends to solve.
#[derive(
    Debug,
    PartialEq,
    Eq,
    Clone,
    Hash,
    PartialOrd,
    Ord,
    strum_macros::AsRefStr,
    strum_macros::Display,
    strum_macros::EnumIter,
    strum_macros::EnumString,
    strum_macros::IntoStaticStr,
    RuleNamespace,
)]
#[repr(u16)]
#[strum(serialize_all = "kebab-case")]
pub enum Category {
    /// Failure to parse a file.
    #[prefix = "E"]
    Error,
    /// Detect code that is bug-prone or likely to be incorrect.
    #[prefix = "C"]
    Correctness,
    /// Obsolescent language features, as determined by the Fortran standard.
    #[prefix = "OB"]
    Obsolescent,
    /// Update to modern Fortran features. Used for less severe issues than `Obsolescent`,
    /// and goes beyond recommendations in the Fortran standard.
    #[prefix = "MOD"]
    Modernisation,
    /// Violation of style conventions.
    #[prefix = "S"]
    Style,
    /// Avoid platform/compiler-specific features.
    #[prefix = "PORT"]
    Portability,
    /// Fortitude specific rules.
    #[prefix = "FORT"]
    Fortitude,
}

pub trait RuleNamespace: Sized {
    /// Returns the prefix that every single code that fortitude uses to identify
    /// rules from this category starts with.
    fn common_prefix(&self) -> &'static str;

    /// Returns the variant corresponding to the given prefix.
    fn from_prefix(preifx: &str) -> Result<Self, &'static str>;

    /// Attempts to parse the given rule code. If the prefix is recognized
    /// returns the respective variant along with the code with the common
    /// prefix stripped.
    fn parse_code(code: &str) -> Option<(Self, &str)>;

    #[allow(dead_code)]
    fn name(&self) -> &'static str;

    #[allow(dead_code)]
    fn description(&self) -> &'static str;
}

pub mod clap_completion {
    use clap::builder::{PossibleValue, TypedValueParser, ValueParserFactory};
    use strum::IntoEnumIterator;

    use crate::registry::Rule;

    #[derive(Clone)]
    pub struct RuleParser;

    impl ValueParserFactory for Rule {
        type Parser = RuleParser;

        fn value_parser() -> Self::Parser {
            RuleParser
        }
    }

    impl TypedValueParser for RuleParser {
        type Value = Rule;

        fn parse_ref(
            &self,
            cmd: &clap::Command,
            arg: Option<&clap::Arg>,
            value: &std::ffi::OsStr,
        ) -> Result<Self::Value, clap::Error> {
            let value = value
                .to_str()
                .ok_or_else(|| clap::Error::new(clap::error::ErrorKind::InvalidUtf8))?;

            Rule::from_code(value).map_err(|_| {
                let mut error =
                    clap::Error::new(clap::error::ErrorKind::ValueValidation).with_cmd(cmd);
                if let Some(arg) = arg {
                    error.insert(
                        clap::error::ContextKind::InvalidArg,
                        clap::error::ContextValue::String(arg.to_string()),
                    );
                }
                error.insert(
                    clap::error::ContextKind::InvalidValue,
                    clap::error::ContextValue::String(value.to_string()),
                );
                error
            })
        }

        fn possible_values(&self) -> Option<Box<dyn Iterator<Item = PossibleValue> + '_>> {
            Some(Box::new(Rule::iter().map(|rule| {
                let name = rule.noqa_code().to_string();
                let help = rule.as_ref().to_string();
                PossibleValue::new(name).help(help)
            })))
        }
    }
}

#[cfg(test)]
mod tests {
    use std::mem::size_of;

    use strum::IntoEnumIterator;

    use super::{Category, Rule, RuleNamespace};

    #[test]
    fn documentation() {
        for rule in Rule::iter() {
            assert!(
                rule.explanation().is_some(),
                "Rule {} is missing documentation",
                rule.as_ref()
            );
        }
    }

    #[test]
    fn check_code_serialization() {
        for rule in Rule::iter() {
            assert!(
                Rule::from_code(&format!("{}", rule.noqa_code())).is_ok(),
                "{rule:?} could not be round-trip serialized."
            );
        }
    }

    #[test]
    fn category_parse_code() {
        for rule in Rule::iter() {
            let code = format!("{}", rule.noqa_code());
            let (category, rest) =
                Category::parse_code(&code).unwrap_or_else(|| panic!("couldn't parse {code:?}"));
            assert_eq!(code, format!("{}{rest}", category.common_prefix()));
        }
    }

    #[test]
    fn rule_size() {
        assert_eq!(2, size_of::<Rule>());
    }
}
