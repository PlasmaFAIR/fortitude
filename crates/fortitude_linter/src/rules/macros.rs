/// Shorthand for `Some(vec![x])`. Note that an empty argument list
/// is forbidden -- functions should just return `None` instead
#[macro_export]
macro_rules! some_vec {
    ($($x:expr),+$(,)?) => {
        Some(vec![$($x),+])
    };
}

#[cfg(test)]
mod tests {
    use pretty_assertions::assert_eq;
    #[test]
    fn test_some_vec() -> anyhow::Result<()> {
        let expected = Some(vec![1, 2, 3]);
        let actual = some_vec![1, 2, 3];

        assert_eq!(expected, actual);
        Ok(())
    }
}
