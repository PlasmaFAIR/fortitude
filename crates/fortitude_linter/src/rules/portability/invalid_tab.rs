use itertools::Itertools;
use ruff_diagnostics::{Diagnostic, Edit, Fix, FixAvailability, Violation};
use ruff_macros::{ViolationMetadata, derive_message_formats};
use ruff_source_file::SourceFile;
use ruff_text_size::{TextRange, TextSize};
use tree_sitter::Node;

/// ## What it does
/// Checks for the use of tab characters as whitespace
///
/// ## Why is this bad?
/// Tabs are not part of the Fortran standard, and compilers may
/// reject the source if using a strict conformance mode (for example,
/// `gfortran -std=f2023 -Werror`).
#[derive(ViolationMetadata)]
pub(crate) struct InvalidTab;

impl Violation for InvalidTab {
    const FIX_AVAILABILITY: FixAvailability = FixAvailability::Sometimes;
    #[derive_message_formats]
    fn message(&self) -> String {
        "Invalid tab character".to_string()
    }

    fn fix_title(&self) -> Option<String> {
        Some("Replace with spaces".to_string())
    }
}

pub fn check_invalid_tab(root: &Node, src: &SourceFile) -> Vec<Diagnostic> {
    src.source_text()
        .char_indices()
        .filter(|(_, c)| *c == '\t')
        .filter(|(index, _)| {
            if let Some(node) = root.named_descendant_for_byte_range(*index, *index) {
                !matches!(node.kind(), "comment" | "string_literal")
            } else {
                false
            }
        })
        .map(|(index, _)| {
            // TODO(peter): This doesn't render particularly well,
            // might be an issue with annotate-snippets?
            let start = TextSize::try_from(index).unwrap();
            let end = start + TextSize::new(1);
            let edit = Edit::replacement("    ".to_string(), start, start + TextSize::new(1));
            Diagnostic::new(InvalidTab, TextRange::new(start, end)).with_fix(Fix::unsafe_edit(edit))
        })
        .collect_vec()
}
