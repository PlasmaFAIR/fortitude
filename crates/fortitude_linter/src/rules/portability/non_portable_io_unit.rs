use crate::ast::FortitudeNode;
use crate::rules::utilities::literal_as_io_unit;
use crate::settings::Settings;
use crate::{AstRule, FromAstNode};
use ruff_diagnostics::{Diagnostic, Violation};
use ruff_macros::{derive_message_formats, ViolationMetadata};
use ruff_source_file::SourceFile;
use tree_sitter::Node;

/// ## What it does
/// Checks for certain literals as units in `read`/`write` statements.
///
/// ## Why is this bad?
/// The Fortran standard does not specify numeric values for `stdin`, `stdout`, or
/// `stderr`, and although many compilers do "pre-connect" units `5`, `6`, and `0`,
/// respectively, some use other numbers. Instead, use the named constants `input_unit`,
/// `output_unit`, or `error_unit` from the `iso_fortran_env` module.
///
/// !!! note
///     An `open` statement with one of these units is completely portable, it is just
///     the use to mean `stdin`/`stdout`/`stderr` without an explicit `open` that is
///     non-portable -- but see also [`magic-io-unit`](magic-io-unit.md) for why it's
///     best to avoid literal integers as IO units altogether.
///
/// ## Options
/// - `check.portability.allow-cray-file-units`
#[derive(ViolationMetadata)]
pub(crate) struct NonPortableIoUnit {
    value: i32,
    kind: String,
    replacement: String,
}

impl Violation for NonPortableIoUnit {
    #[derive_message_formats]
    fn message(&self) -> String {
        let Self { value, kind, .. } = self;
        format!("Non-portable unit '{value}' in '{kind}' statement")
    }

    fn fix_title(&self) -> Option<String> {
        let Self { replacement, .. } = self;
        Some(format!("Use `{replacement}` from `iso_fortran_env`"))
    }
}

impl AstRule for NonPortableIoUnit {
    fn check(settings: &Settings, node: &Node, src: &SourceFile) -> Option<Vec<Diagnostic>> {
        let unit = literal_as_io_unit(node, src)?;

        let value = unit
            .to_text(src.source_text())?
            .parse::<i32>()
            .unwrap_or_default();
        let is_read = node.kind() == "read_statement";
        let is_write = node.kind() == "write_statement";

        let kind = if is_read { "read" } else { "write" }.to_string();

        let mut stdin = vec![5];
        let mut stdout = vec![6];
        let mut stderr = vec![0];

        if !settings.check.portability.allow_cray_file_units {
            stdin.push(100);
            stdout.push(101);
            stderr.push(102);
        }

        let replacement = if is_read && stdin.contains(&value) {
            "input_unit"
        } else if is_write && stdout.contains(&value) {
            "output_unit"
        } else if is_write && stderr.contains(&value) {
            "error_unit"
        } else {
            return None;
        }
        .to_string();

        some_vec!(Diagnostic::from_node(
            Self {
                value,
                kind,
                replacement
            },
            &unit
        ))
    }

    fn entrypoints() -> Vec<&'static str> {
        vec!["read_statement", "write_statement"]
    }
}
