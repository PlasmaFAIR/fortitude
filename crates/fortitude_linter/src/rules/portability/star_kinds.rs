use crate::ast::{dtype_is_plain_number, strip_line_breaks, FortitudeNode};
use crate::settings::Settings;
use crate::{AstRule, FromAstNode};
use ruff_diagnostics::{Diagnostic, Fix, FixAvailability, Violation};
use ruff_macros::{derive_message_formats, ViolationMetadata};
use ruff_source_file::SourceFile;
use tree_sitter::Node;

/// ## What it does
/// Checks for non-standard kind specifiers such as `int*4` or `real*8`
///
/// ## Why is this bad?
/// Types such as 'real*8' or 'integer*4' are not standard Fortran and should be
/// avoided. For these cases, consider instead using 'real(real64)' or
/// 'integer(int32)', where 'real64' and 'int32' may be found in the intrinsic
/// module 'iso_fortran_env'. You may also wish to determine kinds using the
/// built-in functions 'selected_real_kind' and 'selected_int_kind'.
///
/// Fixes to this rule are considered unsafe, as while `dtype*N` is generally
/// understood to mean a `dtype` that occupied `N` bytes, this does not necessarily
/// correspond to `dtype(N)`, which is a `dtype` of 'kind' `N`. For example, the NAG
/// compiler may be configured to use a sequential kind system in which `real*8`
/// corresponds to `real(2)`
///
/// In a future version, we hope to upgrade this to a safe fix by use of parameters
/// in `iso_fortran_env`, as `real*8` should always correspond to `real(real64)`.
#[derive(ViolationMetadata)]
pub(crate) struct StarKind {
    dtype: String,
    size: String,
    kind: String,
}

impl Violation for StarKind {
    const FIX_AVAILABILITY: FixAvailability = FixAvailability::Sometimes;

    #[derive_message_formats]
    fn message(&self) -> String {
        let Self { dtype, size, .. } = self;
        format!("'{dtype}{size}' uses non-standard syntax")
    }

    fn fix_title(&self) -> Option<String> {
        let Self { dtype, kind, .. } = self;
        Some(format!("Replace with '{dtype}({kind})'"))
    }
}

impl AstRule for StarKind {
    fn check(_settings: &Settings, node: &Node, src: &SourceFile) -> Option<Vec<Diagnostic>> {
        let text = src.source_text();
        let dtype = node.child(0)?.to_text(text)?.to_lowercase();
        // TODO: Handle characters
        if !dtype_is_plain_number(dtype.as_str()) {
            return None;
        }
        let kind_node = node.child_by_field_name("kind")?;
        let size = kind_node.to_text(text)?;
        if !size.starts_with('*') {
            return None;
        }

        // Tidy up the kind spec so it's just e.g. '*8'
        let size = strip_line_breaks(size).replace([' ', '\t'], "");

        let literal = kind_node.child_with_name("number_literal")?;
        let kind = literal.to_text(text)?.to_string();
        let replacement = format!("{dtype}({kind})");
        let fix = Fix::unsafe_edit(node.edit_replacement(src, replacement));
        some_vec![Diagnostic::from_node(Self { dtype, size, kind }, &kind_node).with_fix(fix)]
    }

    fn entrypoints() -> Vec<&'static str> {
        vec!["intrinsic_type"]
    }
}
