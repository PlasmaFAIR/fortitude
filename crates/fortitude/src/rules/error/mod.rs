pub mod ioerror;
pub mod syntax_error;

#[cfg(test)]
mod tests {
    use std::convert::AsRef;
    use std::path::Path;

    use anyhow::Result;
    use insta::assert_snapshot;
    use test_case::test_case;

    use crate::apply_common_filters;
    use crate::registry::Rule;
    use crate::settings::Settings;
    use crate::test::test_path;

    #[test_case(Rule::SyntaxError, Path::new("E001.f90"))]
    fn rules(rule_code: Rule, path: &Path) -> Result<()> {
        let snapshot = format!("{}_{}", rule_code.as_ref(), path.to_string_lossy());
        let diagnostics = test_path(
            Path::new("error").join(path).as_path(),
            &[rule_code],
            &Settings::default(),
        )?;
        apply_common_filters!();
        assert_snapshot!(snapshot, diagnostics);
        Ok(())
    }
}
