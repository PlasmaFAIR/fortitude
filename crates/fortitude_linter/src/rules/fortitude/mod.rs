pub(crate) mod allow_comments;

#[cfg(test)]
mod tests {
    use std::convert::AsRef;
    use std::path::Path;

    use anyhow::Result;
    use insta::assert_snapshot;
    use test_case::test_case;

    use crate::apply_common_filters;
    use crate::registry::Rule;
    use crate::settings::Settings;
    use crate::test::test_path;

    #[test_case(Rule::InvalidRuleCodeOrName, Path::new("FORT001.f90"))]
    #[test_case(Rule::UnusedAllowComment, Path::new("FORT002.f90"))]
    #[test_case(Rule::RedirectedAllowComment, Path::new("FORT003.f90"))]
    #[test_case(Rule::DuplicatedAllowComment, Path::new("FORT004.f90"))]
    #[test_case(Rule::DisabledAllowComment, Path::new("FORT005.f90"))]
    fn rules(rule_code: Rule, path: &Path) -> Result<()> {
        let snapshot = format!("{}_{}", rule_code.as_ref(), path.to_string_lossy());
        let diagnostics = test_path(
            Path::new("fortitude").join(path).as_path(),
            &[
                // We enable all these related rules here as we should
                // only enabled exactly one of them
                Rule::InvalidRuleCodeOrName,
                Rule::UnusedAllowComment,
                Rule::RedirectedAllowComment,
                Rule::DuplicatedAllowComment,
                Rule::DisabledAllowComment,
                // and we add this one so we can have something that
                // gets used
                Rule::ImplicitTyping,
            ],
            &Settings::default(),
        )?;
        apply_common_filters!();
        assert_snapshot!(snapshot, diagnostics);
        Ok(())
    }
}
