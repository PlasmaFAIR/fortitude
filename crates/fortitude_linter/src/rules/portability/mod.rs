pub(crate) mod literal_kinds;
pub(crate) mod non_portable_io_unit;
pub(crate) mod star_kinds;

#[cfg(test)]
mod tests {
    use std::convert::AsRef;
    use std::path::Path;

    use anyhow::Result;
    use insta::assert_snapshot;
    use test_case::test_case;

    use crate::apply_common_filters;
    use crate::registry::Rule;
    use crate::rules::portability;
    use crate::settings::{CheckSettings, Settings};
    use crate::test::test_path;

    #[test_case(Rule::NonPortableIoUnit, Path::new("PORT001.f90"))]
    #[test_case(Rule::LiteralKind, Path::new("PORT011.f90"))]
    #[test_case(Rule::LiteralKindSuffix, Path::new("PORT012.f90"))]
    #[test_case(Rule::StarKind, Path::new("PORT021.f90"))]
    fn rules(rule_code: Rule, path: &Path) -> Result<()> {
        let snapshot = format!("{}_{}", rule_code.as_ref(), path.to_string_lossy());
        let diagnostics = test_path(
            Path::new("portability").join(path).as_path(),
            &[rule_code],
            &Settings::default(),
        )?;
        apply_common_filters!();
        assert_snapshot!(snapshot, diagnostics);
        Ok(())
    }

    #[test]
    fn warn_port001_no_cray_units() -> Result<()> {
        let rule_code = Rule::NonPortableIoUnit;
        let path = Path::new("PORT001.f90");
        let snapshot = format!(
            "{}_{}_no_cray_units",
            rule_code.as_ref(),
            path.to_string_lossy()
        );
        let default = Settings::default();
        #[allow(clippy::needless_update)]
        let settings = Settings {
            check: CheckSettings {
                portability: portability::settings::Settings {
                    allow_cray_file_units: true,
                },
                ..default.check
            },
            ..default
        };
        let diagnostics = test_path(
            Path::new("portability").join(path).as_path(),
            &[rule_code],
            &settings,
        )?;
        apply_common_filters!();
        assert_snapshot!(snapshot, diagnostics);
        Ok(())
    }
}

pub mod settings {
    use crate::display_settings;
    use ruff_macros::CacheKey;
    use std::fmt::{Display, Formatter};

    #[derive(Debug, Clone, Default, CacheKey)]
    pub struct Settings {
        pub allow_cray_file_units: bool,
    }

    impl Display for Settings {
        fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
            display_settings! {
                formatter = f,
                namespace = "check.portability",
                fields = [self.allow_cray_file_units]
            }
            Ok(())
        }
    }
}
