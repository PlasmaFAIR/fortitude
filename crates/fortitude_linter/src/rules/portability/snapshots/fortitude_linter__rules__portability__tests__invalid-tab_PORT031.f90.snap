---
source: crates/fortitude_linter/src/rules/portability/mod.rs
expression: diagnostics
snapshot_kind: text
---
./resources/test/fixtures/portability/PORT031.f90:4:1: PORT031 [*] Invalid tab character
  |
2 | ! the horrible whitespace!
3 | program invalid_tab
4 |     implicit none
  | ^^^^ PORT031
5 |     print*, "Don't flag this:    "
6 |     ! Or this:|    |
  |
  = help: Replace with spaces

ℹ Unsafe fix
1 1 | ! If you touch this flie, be careful your editor doesn't auto-correct
2 2 | ! the horrible whitespace!
3 3 | program invalid_tab
4   |-	implicit none
  4 |+    implicit none
5 5 |     print*, "Don't flag this:	"
6 6 |     ! Or this:|	|
7 7 |     if	(.true.)	then

./resources/test/fixtures/portability/PORT031.f90:7:7: PORT031 [*] Invalid tab character
  |
5 |     print*, "Don't flag this:    "
6 |     ! Or this:|    |
7 |     if    (.true.)    then
  |       ^^^^ PORT031
8 |         print*, "Mixed tab/space"
9 |     else
  |
  = help: Replace with spaces

ℹ Unsafe fix
4 4 | 	implicit none
5 5 |     print*, "Don't flag this:	"
6 6 |     ! Or this:|	|
7   |-    if	(.true.)	then
  7 |+    if    (.true.)	then
8 8 |     	print*, "Mixed tab/space"
9 9 |     else
10 10 | 		print*, "Two tabs"

./resources/test/fixtures/portability/PORT031.f90:7:16: PORT031 [*] Invalid tab character
  |
5 |     print*, "Don't flag this:    "
6 |     ! Or this:|    |
7 |     if    (.true.)    then
  |                   ^^^^ PORT031
8 |         print*, "Mixed tab/space"
9 |     else
  |
  = help: Replace with spaces

ℹ Unsafe fix
4 4 | 	implicit none
5 5 |     print*, "Don't flag this:	"
6 6 |     ! Or this:|	|
7   |-    if	(.true.)	then
  7 |+    if	(.true.)    then
8 8 |     	print*, "Mixed tab/space"
9 9 |     else
10 10 | 		print*, "Two tabs"

./resources/test/fixtures/portability/PORT031.f90:8:5: PORT031 [*] Invalid tab character
   |
 6 |     ! Or this:|    |
 7 |     if    (.true.)    then
 8 |         print*, "Mixed tab/space"
   |     ^^^^ PORT031
 9 |     else
10 |         print*, "Two tabs"
   |
   = help: Replace with spaces

ℹ Unsafe fix
5 5 |     print*, "Don't flag this:	"
6 6 |     ! Or this:|	|
7 7 |     if	(.true.)	then
8   |-    	print*, "Mixed tab/space"
  8 |+        print*, "Mixed tab/space"
9 9 |     else
10 10 | 		print*, "Two tabs"
11 11 |     end if

./resources/test/fixtures/portability/PORT031.f90:10:1: PORT031 [*] Invalid tab character
   |
 8 |         print*, "Mixed tab/space"
 9 |     else
10 |         print*, "Two tabs"
   | ^^^^ PORT031
11 |     end if
12 | end program invalid_tab
   |
   = help: Replace with spaces

ℹ Unsafe fix
7  7  |     if	(.true.)	then
8  8  |     	print*, "Mixed tab/space"
9  9  |     else
10    |-		print*, "Two tabs"
   10 |+    	print*, "Two tabs"
11 11 |     end if
12 12 | end program invalid_tab

./resources/test/fixtures/portability/PORT031.f90:10:2: PORT031 [*] Invalid tab character
   |
 8 |         print*, "Mixed tab/space"
 9 |     else
10 |         print*, "Two tabs"
   |     ^^^^ PORT031
11 |     end if
12 | end program invalid_tab
   |
   = help: Replace with spaces

ℹ Unsafe fix
7  7  |     if	(.true.)	then
8  8  |     	print*, "Mixed tab/space"
9  9  |     else
10    |-		print*, "Two tabs"
   10 |+	    print*, "Two tabs"
11 11 |     end if
12 12 | end program invalid_tab
